File: dart_client/lib/src/services/collections_service.dart
import 'package:dio/dio.dart';
import '../dtos.dart';
import '../enums.dart';

class CollectionsService {
  final Dio _dio;
  CollectionsService(this._dio);

  Future<DTOResponse<CollectionListResponseDTO>> getCollections({}) async {
    final response = await _dio.get('/collections');
    return DTOResponse<CollectionListResponseDTO>.fromJson(response.data, (json) => CollectionListResponseDTO.fromJson(json as Map<String, dynamic>));
  }

  Future<DTOResponse<IdDataResponseDTO>> postCollections({required CreateCollectionRequestDTO body}) async {
    final response = await _dio.post('/collections', data: body.toJson());
    return DTOResponse<IdDataResponseDTO>.fromJson(response.data, (json) => IdDataResponseDTO.fromJson(json as Map<String, dynamic>));
  }

  Future<DTOResponse<CollectionDetailResponseDTO>> getCollectionsId({required int id}) async {
    final response = await _dio.get('/collections/$id');
    return DTOResponse<CollectionDetailResponseDTO>.fromJson(response.data, (json) => CollectionDetailResponseDTO.fromJson(json as Map<String, dynamic>));
  }

  Future<DTOResponse<GenericResponseDTO>> putCollectionsId({required UpdateCollectionRequestDTO body, required int id}) async {
    final response = await _dio.put('/collections/$id', data: body.toJson());
    return DTOResponse<GenericResponseDTO>.fromJson(response.data, (json) => GenericResponseDTO.fromJson(json as Map<String, dynamic>));
  }

  Future<DTOResponse<GenericResponseDTO>> deleteCollectionsId({required int id}) async {
    final response = await _dio.delete('/collections/$id');
    return DTOResponse<GenericResponseDTO>.fromJson(response.data, (json) => GenericResponseDTO.fromJson(json as Map<String, dynamic>));
  }

}

File: js_client/contracts/services/collections_service.js
import { DTOResponse, DTOParams, EmptyParamsDTO, SearchQuestionsParamsDTO, GetQuestionParamsDTO, GetQuestionsBatchParamsDTO, UpdateQuestionParamsDTO, DeleteQuestionParamsDTO, UpdateCollectionParamsDTO, EmptyRequestDTO, BaseResponse, LoginResponseDTO, GenericResponseDTO, MessageResponseDTO, MustChangePasswordResponseDTO, UserStatusResponseDTO, ProfileDataResponseDTO, QuestionListResponseDTO, QuestionDataResponseDTO, CollectionListResponseDTO, IdDataResponseDTO, CollectionDetailResponseDTO, MarkListResponseDTO, FriendRequestResponseDTO, UserListResponseDTO, LoginRequestDTO, LoginDataDTO, SignupRequestDTO, MessageDataDTO, RecoverPasswordRequestDTO, VerifyEmailRequestDTO, MustChangePasswordDataDTO, ChangePasswordRequestDTO, UserStatusDataDTO, RegisterPushTokenRequestDTO, ProfileDataDTO, UpdateProfileRequestDTO, AnswerDataDTO, QuestionDataDTO, LocalizationDTO, CreateAnswerInputDTO, CreateQuestionInputDTO, CreateQuestionsRequestDTO, UpdateAnswerInputDTO, UpdateQuestionRequestDTO, CollectionDataDTO, CollectionDetailDataDTO, CreateCollectionRequestDTO, IdDataDTO, UpdateCollectionRequestDTO, MarkDataDTO, SendFriendRequestDTO, FriendRequestDataDTO, RespondFriendRequestRequestDTO, FriendRequestListResponseDTO, PublicUserProfileDTO, ErrorDetailsDTO } from '../dtos.js';
import * as Enums from '../enums.js';

export class CollectionsService {
  constructor(baseUrl, fetcher) {
    this.baseUrl = baseUrl;
    this.fetcher = fetcher;
  }

  async getCollections({  } = {}) {
    const url = new URL(`${this.baseUrl}/collections`);
    const headers = { 'Content-Type': 'application/json' };
    const response = await this.fetcher(url.toString(), {
      method: 'GET',
      headers,
    });
    const json = await response.json();
    return DTOResponse.fromJson(json, (data) => CollectionListResponseDTO.fromJson(data));
  }

  async postCollections({ body, bearerToken } = {}) {
    const url = new URL(`${this.baseUrl}/collections`);
    const headers = { 'Content-Type': 'application/json' };
    if (bearerToken) headers['Authorization'] = `Bearer ${bearerToken}`;
    const response = await this.fetcher(url.toString(), {
      method: 'POST',
      headers,
      body: JSON.stringify(body.toJson())
    });
    const json = await response.json();
    return DTOResponse.fromJson(json, (data) => IdDataResponseDTO.fromJson(data));
  }

  async getCollectionsId({ id,  } = {}) {
    const url = new URL(`${this.baseUrl}/collections/${encodeURIComponent(id)}`);
    const headers = { 'Content-Type': 'application/json' };
    const response = await this.fetcher(url.toString(), {
      method: 'GET',
      headers,
    });
    const json = await response.json();
    return DTOResponse.fromJson(json, (data) => CollectionDetailResponseDTO.fromJson(data));
  }

  async putCollectionsId({ body, id, bearerToken } = {}) {
    const url = new URL(`${this.baseUrl}/collections/${encodeURIComponent(id)}`);
    const headers = { 'Content-Type': 'application/json' };
    if (bearerToken) headers['Authorization'] = `Bearer ${bearerToken}`;
    const response = await this.fetcher(url.toString(), {
      method: 'PUT',
      headers,
      body: JSON.stringify(body.toJson())
    });
    const json = await response.json();
    return DTOResponse.fromJson(json, (data) => GenericResponseDTO.fromJson(data));
  }

  async deleteCollectionsId({ id, bearerToken } = {}) {
    const url = new URL(`${this.baseUrl}/collections/${encodeURIComponent(id)}`);
    const headers = { 'Content-Type': 'application/json' };
    if (bearerToken) headers['Authorization'] = `Bearer ${bearerToken}`;
    const response = await this.fetcher(url.toString(), {
      method: 'DELETE',
      headers,
    });
    const json = await response.json();
    return DTOResponse.fromJson(json, (data) => GenericResponseDTO.fromJson(data));
  }

}
